builtin/providers/google/data_source_storage_object_signed_url.go

```diff
@@ -0,0 +1,297 @@
+package google
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"github.com/hashicorp/terraform/helper/pathorcontents"
+	"github.com/hashicorp/terraform/helper/schema"
+	"golang.org/x/oauth2/google"
+	"golang.org/x/oauth2/jwt"
+	"log"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const gcsBaseUrl = "https://storage.googleapis.com"
+const googleCredentialsEnvVar = "GOOGLE_APPLICATION_CREDENTIALS"
+
+func dataSourceGoogleSignedUrl() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGoogleSignedUrlRead,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			//TODO: implement support
+			//"content_type": &schema.Schema{
+			//	Type:     schema.TypeString,
+			//	Optional: true,
+			//	Default:  "",
+			//},
+			"credentials": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"duration": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "1h",
+			},
+			"http_method": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "GET",
+			},
+			//TODO: implement support
+			//"http_headers": &schema.Schema{
+			//	Type:     schema.TypeList,
+			//	Optional: true,
+			//},
+			//TODO: implement support
+			//"md5_digest": &schema.Schema{
+			//	Type:     schema.TypeString,
+			//	Optional: true,
+			//	Default:  "",
+			//},
+			"path": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"signed_url": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceGoogleSignedUrlRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	// Build UrlData object from data source attributes
+	urlData := &UrlData{}
+
+	// HTTP Method
+	if method, ok := d.GetOk("http_method"); ok && len(method.(string)) >= 3 {
+		urlData.HttpMethod = method.(string)
+	} else {
+		return fmt.Errorf("not a valid http method")
+	}
+
+	// convert duration to an expiration datetime (unix time in seconds)
+	durationString := "1h"
+	if v, ok := d.GetOk("duration"); ok {
+		durationString = v.(string)
+	}
+	duration, err := time.ParseDuration(durationString)
+	if err != nil {
+		return fmt.Errorf("could not parse duration")
+	}
+	expires := time.Now().Unix() + int64(duration.Seconds())
+	urlData.Expires = int(expires)
+
+	// object path
+	path := []string{
+		"",
+		d.Get("bucket").(string),
+		d.Get("path").(string),
+	}
+	objectPath := strings.Join(path, "/")
+	urlData.Path = objectPath
+
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
Can we remove these separator lines? The comment text should be sufficient to delineate the various sections of code.